library/std/src/sys/pal/unix/pi_futex.rs

```diff
@@ -0,0 +1,171 @@
+#[cfg(any(target_os = "linux", target_os = "android"))]
+mod linux {
+    use crate::ops::Deref;
+    use crate::sync::atomic::AtomicU32;
+    use crate::sys::cvt;
+    use crate::{io, ptr};
+
+    pub type State = u32;
+
+    pub struct Futex(AtomicU32);
+
+    impl Futex {
+        pub const fn new() -> Futex {
+            Futex(AtomicU32::new(0))
+        }
+    }
+
+    impl Deref for Futex {
+        type Target = AtomicU32;
+        fn deref(&self) -> &AtomicU32 {
+            &self.0
+        }
+    }
+
+    pub const fn unlocked() -> State {
+        0
+    }
+
+    pub fn locked() -> State {
+        (unsafe { libc::gettid() }) as _
```

Comment:
We can store it in the `Thread` structure.

We could additionally install a `MADV_WIPEONFORK` page containing an atomic function pointer that resolves either to "get the cached result" or "do a syscall" functions... though that requires 4.14. An `atfork` handler would work too but adding one might have other exciting consequences.

Reply:
> We cannot use atfork. e.g. clone and clone3 do not call atfork handlers. We cannot assume the programmers always use syscall wrappers from glibc.

libc authors have said that not using the wrappers and then calling libc functions after the clone is UB. Since the standard library on those targets relies on rust that's also Rust UB.

> Somewhat complicated. I can implement this, but I'm afraid it's not zero-cost (we need one memory page per thread to store tid.)

I don't think so. A shared flag or a sort of generation marker for all threads should be sufficient.

Reply:
Thread id is cached in thread-local storage in the latest commit. Would you plz review it?